File: web/src/lib/explain.js
// @ts-check
/**
 * The entire "plain English" system.
 *
 * One generic template, no per-algorithm code (invariant I2), no language
 * model. Everything it needs is already in the event: `expr` says how the value
 * was computed, `deps` says what was read AND what those cells held at read
 * time.
 *
 * Those dep values are SNAPSHOTTED into the event rather than looked up during
 * replay. That is what makes the explanation exact while stepping BACKWARD:
 * current state during a rewind is not the state the read saw. ADR 0005.
 *
 * The Go twin lives in cmd/orrery/explain.go and drives the terminal player.
 * Two consumers, one set of rules, zero algorithm-specific text anywhere.
 */

import { addrLabel, fmtValue } from './value.js';

/**
 * @typedef {object} Explanation
 * @property {'set'|'call'|'ret'|'init'|'none'} kind
 * @property {string} lead
 * @property {string} because
 * @property {Array<{label:string, value:string, s:string, at:Array}>} where
 * @property {number} ln
 */

/**
 * @param {Array<object>} events the events of the current step
 * @param {import('../player/prepass.js').Index} [index]
 * @returns {Explanation}
 */
export function explain(events, index) {
  const empty = { kind: /** @type {const} */ ('none'), lead: '', because: '', where: [], ln: 0 };
  if (!events || events.length === 0) return empty;

  // A grouped step (a swap) has several events. Lead with the group's note and
  // merge the provenance -- describing only the first write would be a lie
  // about what the step did.
  const primary = events.find((e) => e.t === 'set') ?? events[0];
  const note = events.find((e) => e.note)?.note ?? '';

  const where = [];
  for (const e of events) {
    for (const d of e.deps ?? []) {
      where.push({
        label: addrLabel(d.s, d.at ?? []),
        value: fmtValue(d.v),
        s: d.s,
        at: d.at ?? [],
      });
    }
  }

  switch (primary.t) {
    case 'init':
      return { ...empty, kind: 'init', lead: `${primary.s} created`, ln: primary.ln ?? 0 };

    case 'call': {
      const args = (primary.args ?? []).map((a) => `${a.n}=${fmtValue(a.v)}`).join(', ');
      return {
        kind: 'call', lead: `${primary.fn}(${args})`,
        because: note, where, ln: primary.ln ?? 0,
      };
    }

    case 'ret': {
      let because = note || primary.expr || '';
      // "already computed at step N" comes from the pre-pass firstWrite index,
      // not from the event -- the trace records what was read, and the app
      // figures out when it was written.
      const d = (primary.deps ?? [])[0];
      if (index && d) {
        const ev = index.firstWrite.get(`${d.s} ${(d.at ?? []).join('/')}`);
        if (ev !== undefined) {
          const k = index.steps.findIndex((s) => ev >= s.e0 && ev < s.e1);
          if (k >= 0) because = because ? `${because} — computed at step ${k + 1}` : `computed at step ${k + 1}`;
        }
      }
      return {
        kind: 'ret', lead: `returns ${fmtValue(primary.v)}`,
        because, where, ln: primary.ln ?? 0,
      };
    }

    case 'set': {
      const writes = events.filter((e) => e.t === 'set');
      let lead;
      if (writes.length > 1) {
        lead = writes.map((e) => `${addrLabel(e.s, e.at ?? [])} → ${fmtValue(e.to)}`).join('  ·  ');
      } else {
        const a = addrLabel(primary.s, primary.at ?? []);
        lead = primary.from === null || primary.from === undefined
          ? `${a} ← ${fmtValue(primary.to)}`
          : `${a}: ${fmtValue(primary.from)} → ${fmtValue(primary.to)}`;
      }
      return {
        kind: 'set', lead,
        because: note || primary.expr || '',
        where, ln: primary.ln ?? 0,
      };
    }
  }
  return empty;
}

File: cmd/orrery/explain.go
package main

import (
	"fmt"
	"strings"

	"github.com/sanu1001/orrery/internal/trace"
)

// explain is the Go twin of web/src/lib/explain.js.
//
// This is the ENTIRE "plain English" system: one generic template, no
// per-algorithm code (invariant I2), no language model. Everything it needs is
// already in the event -- `expr` says how the value was computed, `deps` says
// what was read and what those cells held AT READ TIME.
//
// Snapshotting the dep values into the event rather than looking them up during
// replay is what makes the explanation exact while stepping BACKWARD: current
// state during a rewind is not the state the read saw. ADR 0005.
func explain(e trace.Event, t *trace.Trace) string {
	switch e.T {
	case trace.Init:
		return ""
	case trace.Call:
		return callLabel(e)
	case trace.Ret:
		var b strings.Builder
		fmt.Fprintf(&b, "returns %s", trace.Canon(e.V))
		if e.Note != "" {
			fmt.Fprintf(&b, " -- %s", e.Note)
		} else if e.Expr != "" {
			fmt.Fprintf(&b, " -- %s", e.Expr)
		}
		if w := where(e.Deps); w != "" {
			fmt.Fprintf(&b, ", %s", w)
		}
		return b.String()
	case trace.Set:
		addr := addrLabel(e.S, e.At)
		var b strings.Builder
		if e.From == nil && e.To != nil {
			fmt.Fprintf(&b, "%s <- %s", addr, trace.Canon(e.To))
		} else {
			fmt.Fprintf(&b, "%s: %s -> %s", addr, trace.Canon(e.From), trace.Canon(e.To))
		}
		switch {
		case e.Note != "":
			fmt.Fprintf(&b, " -- %s", e.Note)
		case e.Expr != "":
			fmt.Fprintf(&b, " -- %s", e.Expr)
		}
		if w := where(e.Deps); w != "" {
			fmt.Fprintf(&b, ", %s", w)
		}
		return b.String()
	}
	return ""
}

// addrLabel renders "dp[3][4]" for numeric paths and "L.n3.next" for named
// ones, because those are how a reader of the ALGORITHM would write them.
func addrLabel(name string, at trace.Path) string {
	if len(at) == 0 {
		return name
	}
	numeric := true
	for _, s := range at {
		if !s.IsIdx {
			numeric = false
			break
		}
	}
	var b strings.Builder
	b.WriteString(name)
	if numeric {
		for _, s := range at {
			fmt.Fprintf(&b, "[%d]", s.I)
		}
		return b.String()
	}
	for _, s := range at {
		if s.IsIdx {
			fmt.Fprintf(&b, "[%d]", s.I)
		} else {
			fmt.Fprintf(&b, ".%s", s.S)
		}
	}
	return b.String()
}

func where(deps []trace.Dep) string {
	if len(deps) == 0 {
		return ""
	}
	parts := make([]string, 0, len(deps))
	for _, d := range deps {
		parts = append(parts, fmt.Sprintf("%s was %s", addrLabel(d.S, d.At), trace.Canon(d.V)))
	}
	return "where " + strings.Join(parts, ", ")
}
